ui/org.eclipse.pde.core/src/org/eclipse/pde/internal/core/project/BundleProjectService.java

```diff
@@ -89,110 +89,94 @@ public static synchronized IBundleProjectService getDefault() {
-	 * modify a project.
-	 *
-	 * @param project project
-	 * @return bundle description for the associated project
-	 * @exception CoreException if unable to create a description on an existing project
-	 */
 	@Override
 	public IBundleProjectDescription getDescription(IProject project) throws CoreException {
 		return new BundleProjectDescription(project);
 	}
 
-	/**
-	 * Creates and returns a new host description.
-	 *
-	 * @param name symbolic name of the host
-	 * @param range version constraint or <code>null</code>
-	 * @return host description
-	 */
 	@Override
 	public IHostDescription newHost(String name, VersionRange range) {
-		return new HostDescriptoin(name, range);
+		return new HostDescription(name, range);
+	}
+
+	record HostDescription(String name, VersionRange version) implements IHostDescription {
+		@Override
+		public String getName() {
+			return name();
+		}
+
+		@Override
+		public VersionRange getVersion() {
+			return getVersion();
+		}
 	}
 
-	/**
-	 * Creates and returns a new package import description.
-	 *
-	 * @param name fully qualified name of imported package
-	 * @param range version constraint or <code>null</code>
-	 * @param optional whether the import is optional
-	 * @return package import description
-	 */
 	@Override
 	public IPackageImportDescription newPackageImport(String name, VersionRange range, boolean optional) {
 		return new PackageImportDescription(name, range, optional);
 	}
 
-	/**
-	 * Constructs a new package export description.
-	 *
-	 * @param name fully qualified package name
-	 * @param version version or <code>null</code>
-	 * @param api whether the package is considered API
-	 * @param friends symbolic names of bundles that are friends, or <code>null</code>; when
-	 *  friends are specified the package will not be API
-	 * @return package export description
-	 */
+	record PackageImportDescription(String name, VersionRange version, boolean isOptional)
+			implements IPackageImportDescription {
+		@Override
+		public String getName() {
+			return name();
+		}
+
+		@Override
+		public VersionRange getVersion() {
+			return version();
+		}
+	}
+
 	@Override
 	public IPackageExportDescription newPackageExport(String name, Version version, boolean api, String[] friends) {
-		return new PackageExportDescription(name, version, friends, api);
+		List<String> friendsList = friends != null ? List.of(friends) : List.of();
+		return new PackageExportDescription(name, version, friendsList, friendsList.isEmpty() ? api : false);
+	}
+
+	record PackageExportDescription(String name, Version version, List<String> friends, boolean isApi)
+			implements IPackageExportDescription {
+
+		@Override
+		public String getName() {
+			return name();
+		}
+
+		@Override
+		public Version getVersion() {
+			return version();
+		}
+
+		@Override
+		public String[] getFriends() {
+			return friends.isEmpty() ? null : friends.toArray(String[]::new);
```

Comment:
Is it possible to avoid returning `null` for array?

Reply:
That does seem better.

Reply:
> Is it possible to avoid returning `null` for array?

Unfortunately not really because the doc of `IPackageExportDescription.getFriends()` states:
```
Returns the declared friends of this package or <code>null</code> if none.
```

But in the API rework in https://github.com/eclipse-pde/eclipse.pde/pull/1341, the replacement of that method returns an empty `(Sorted)Set` if an exported package has no friends.

In general this change is intended as clean-up without behavioral change and to keep the change in #1341 smaller. But I'm happy to discuss any enhancements of this API there.

Reply:
I'm submitting this now, lets continue any API discussion in #1341.